[WIP] Xcode 8 beta 4 and 5 + Swift 3 compatibility

mokagio opened the pull request:
Hi 👋. I had a go at adding Xcode 8 beta 4 support and this is as far as I've
got.

There are still two issues:

- `Boolean` has been removed in a breaking way, [SE-109](https://github.com/apple/swift-evolution/blob/a76560fbd834c3a4b0e801f010f06e26e8387073/proposals/0109-remove-boolean.md), which might require us to change the behaviour of `beTruthy` and `beFalsy`.
- The test `testNegativeMatchesWithClosure` is failing <del>due to `if errorType != nil || closure != nil` inside `setFailureMessageForError` resulting false despite `closure` being not `nil` 😳</del>

**It is definitely possible that I'm missing something here, if so feel free to cherry pick only the useful changes.**

### More on the `Boolean` issue

As said `Boolean` has been removed, making this code fail to build.

```swift
public func beTruthy<T>() -> MatcherFunc<T> {
    return MatcherFunc { actualExpression, failureMessage in
        failureMessage.postfixMessage = "be truthy"
        let actualValue = try actualExpression.evaluate()
        if let actualValue = actualValue {
            if let actualValue = actualValue as? Boolean {
                return actualValue.boolValue == true
            }
        }
        return actualValue != nil
    }
}
```

I was able to work around it by simply using `Bool`:

```diff
- if let actualValue = actualValue as? Boolean {
-     return actualValue.boolValue == true
+ if let actualValue = actualValue as? Bool {
+     return actualValue == true
}
```

This change though makes the test using `ConvertsToBool`, which I tried to tweak to make it conform to `ExpressibleByBooleanLiteral` in 72785699c3050c5c8cf3967b52de89fbcc021edc, fail.

That is due to the fact that `ExpressibleByBooleanLiteral`, as far as I've seen, only defines a way to init a value using a `BooleanLiteral`, which is a `Bool` typealias, not how to convert it.

`ExpressibleByBooleanLiteral` has been introduced by [SE-115](https://github.com/apple/swift-evolution/blob/57b4754fadad9dcea55a4a4f003691a224048090/proposals/0115-literal-syntax-protocols.md).

And even if it did have the `boolValue` method, since the protocol is generic, we wouldn't be able to use it in an `as` clause.

mokagio commented on Tests/Nimble/Matchers/BeLogicalTest.swift:
```diff
@@ -1,9 +1,18 @@
+        case false: self = .falseLike
+        }
+    }
+
```
When I implemented these I though that `ExpressibleByBooleanLiteral` came with `boolValue`, and that doesn't seem to be the case.

These changes are probably useless for the goal of making this type able to treated as a boolean.

I'd be happy to remove this commit if this is the case, but I decided to leave for the sake of providing a better picture of what I think is the issue.

fbartho commented:
Can't you implement beTruthy multiple times for numbers, bool, Optionals? Then overload-resolution will pick the correct func? -- Implement beFalsy in terms of beTruthy, or vice-versa.

norio-nomura commented:
I think it would be better to defining `Boolean` (`BooleanProtocol`?) protocol in Nimble instead of using overload.

mokagio commented:
> I think it would be better to defining `Boolean` (`BooleanProtocol`?) protocol in Nimble instead of using overload.

I wouldn't be keen on the overload option either.

I wonder if there actually is a _native_ Swift protocol that types can conform to and declare `boolValue` 🤔

That would be, in my opinion, the best solution.

Correct me if I'm wrong, but defining a `BooleanProtocol` within Nimble would mean that code under test would need to `import Nimble` in order to conform to the protocol, and that doesn't sound nice. 

So far Nimble has been able to assert code written without knowing about it, and I feel like this is the only way to write a good matcher framework.

fbartho commented:
Why aren't you keen on the overload option? -- Just to be clear, we all agree that the overload option would be pure native Swift, just concrete types rather than single generic implementation -- right?

Alternately, you can add a NimbleBoolean protocol, and implement extensions on all of the Standard library Truthy-types so that they conform to NimbleBoolean. Then library consumers would only need to specialize if they're adding more Truthy things.

norio-nomura commented:
We can define protocol as *fileprivate* in `BeLogical.swift`.
```swift
fileprivate protocol Boolean {
    var boolValue: Bool { get }
}

extension Bool : Boolean {
    var boolValue: Bool {
        return self
    }
}

extension ObjCBool: Boolean {}
extension NSNumber: Boolean {}
extension NSString: Boolean {}
```
I don't know how many standard types need to be declared conforming.
I believe using protocol will be better than use overload.
I can't say which is better whether the protocol should be exposed from Nimble or not.
If we won't do, we should remove `testShouldMatchBoolConvertibleTypesThatConvertToTrue` and `testShouldNotMatchBoolConvertibleTypesThatConvertToFalse`.


I guess Nimble does not have any tests that depends on standard types conforming to `Boolean`.
So, I think following:
```diff
- if let actualValue = actualValue as? Boolean {
-     return actualValue.boolValue == true
+ if let actualValue = actualValue as? Bool {
+     return actualValue == true
}
```
without declaring `Boolean` protocol and removing `*BoolConvertible*` tests would be an another choice.

I don't have personal preference about that.

norio-nomura commented:
Another option using `ExpressibleByBooleanLiteral & Equatable` instead of defining protocol:
```swift
public func beTruthy<T: ExpressibleByBooleanLiteral & Equatable>() -> MatcherFunc<T> {
    return MatcherFunc { actualExpression, failureMessage in
        failureMessage.postfixMessage = "be truthy"
        let actualValue = try actualExpression.evaluate()
        if let actualValue = actualValue {
            return actualValue == (true as T)
        }
        return actualValue != nil
    }
}
```

mokagio commented:
@norio-nomura your [suggested implementation with `ExpressibleByBoolenaLiteral & Equeatable`](https://github.com/norio-nomura/Nimble/pull/5#issuecomment-237480994) solves the issues I'm having but results in two new ones:

<img width="1061" alt="screen shot 2016-08-05 at 9 37 18 am" src="https://cloud.githubusercontent.com/assets/1218433/17421920/e87422d4-5af0-11e6-8ad5-8e86e9a3dfc7.png">

<img width="1058" alt="screen shot 2016-08-05 at 9 37 00 am" src="https://cloud.githubusercontent.com/assets/1218433/17421928/f4a20044-5af0-11e6-9ba7-9b306008d344.png">

I was keen on that approach because it would have been unobtrusive and by using only protocols portable.

I think the issue is that `ExpressibleByBooleanLiteral` is meant to be used to represent types that can be _initialized_ using a boolean value. The header says:

> A type that can be initialized with the Boolean literals `true` and
> `false`.
>
> Only three types provided by Swift---`Bool`, `DarwinBoolean`, and
> `ObjCBool`---are treated as Boolean values. Expanding this set to include
> types that represent more than simple Boolean values is discouraged.
>
> To add `ExpressibleByBooleanLiteral` conformance to your custom type,
> implement the `init(booleanLiteral:)` initializer that creates an instance
> of your type with the given Boolean value.

I can find a protocol that types can conform to to declare they can be represented as boolean values.

If such protocol doesn't exist, then the only options I see are:

- Overloading using `fileprivate` as suggested [here](https://github.com/norio-nomura/Nimble/pull/5#issuecomment-237479436).
- Remove `testShouldNotMatchBoolConvertibleTypesThatConvertToFalse` and `testShouldMatchBoolConvertibleTypesThatConvertToTrue`.

norio-nomura commented:
@mokagio I guess the error is bug because `NSNumber` conforms `ExpressibleByBooleanLiteral` and the error message is:
> 2016-08-05 21:13:51.899 xctest[68742:1689305] +[NSNumber initWithBooleanLiteral:]: unrecognized selector sent to class 0x101c3a278

norio-nomura commented:
filed [SR-2290](https://bugs.swift.org/browse/SR-2290)

mokagio commented:
@norio-nomura I've incorporated your `ExpressibleByBooleanLiteral & Equatable` suggestion in the PR, and added a `FIXME` workaround for `NSNumber`.

Now only one test left failing.

Would be cool to have this sorted before beta 5, eh-eh 😅

norio-nomura commented:
@mokagio Could you please check running tests of Nimble-OSX on your mac?

mokagio commented:
@norio-nomura wooops... totally forgot about OS X 😳😅

mokagio commented on Tests/Nimble/AsynchronousTest.swift:
```diff
@@ -63,15 +63,26 @@ final class AsyncTest: XCTestCase, XCTestCaseProvider {
+
+        var asyncOperation: () -> () = { sleepThenSetValueTo(1) }
+
+        if #available(OSX 10.10, *) {
```
Another option would be to drop support for version of OSX < 10.10.

I don't develop for Mac so I don't know if that would make sense or not... I guess an issue could be opened on the Nimble trunk and have a proper discussion there.

mokagio commented:
4e52103 should be squashed onto c3fe2c5 before merging

mokagio commented:
I've been looking into the [only remaining test failure](https://github.com/mokagio/Nimble/blob/mokagio/xcode-8-beta-4/Tests/Nimble/Matchers/ThrowErrorTest.swift#L147)

```swift
let innerFailureMessage = "expected to equal <foo>, got <\(moduleName).Error>"
let closure = { (error: Error) in
    print("** In closure! With domain \(error._domain)")
    expect(error._domain).to(equal("foo"))
}

failsWithErrorMessage([innerFailureMessage, "expected to throw error from type <Error> that satisfies block, got <Laugh>"]) {
    expect { throw Error.Laugh }.to(throwError(closure: closure))
}
```

The failure message is:

> test failure: -[ThrowErrorTest testNegativeMatchesWithClosure()] failed: 
> Got failure message:
> "expected to throw error that satisfies block, got <laugh>",
> but expected:
> "expected to throw error from type <NimbleError> that satisfies block, got <laugh>"

There is only one function in the whole framework that can append "from type ..." to the `postfixMessage`. I set some breakpoints and saw that function is not called at all.

I compared the behaviour of the test when run on trunk Nimble on Xcode 7 (Swift 2.2 obviously) and this PR on Xcode 8 beta 5 Swift 3.

The Swift 2.2 calls the [generic version of `throwError(closure: closure)`](https://github.com/mokagio/Nimble/blob/mokagio/xcode-8-beta-4/Sources/Nimble/Matchers/ThrowError.swift#L3), while Swift 3 calls the [non generic one, specialised with the closure parameter only](https://github.com/mokagio/Nimble/blob/mokagio/xcode-8-beta-4/Sources/Nimble/Matchers/ThrowError.swift#L32).

It should not surprise to find out that the non-generic version doesn't call the method to set the failure message by adding error type information to it.

I am not sure which of the two behaviours is correct, the 2.2 or the 3 one... They both makes sense, and I don't know how Swift is supposed to resolve these kind of overloading.

---

This is an idea to _fix_ the failing test.

The description of the `throwError` which is now called by Swift 3 is:

> A Nimble matcher that succeeds when the actual expression throws any error or when the passed closures' arbitrary custom matching succeeds.

There is no mention to the type of the error in the description of the matchers behaviour.

Moreover when I read an expression like

```swift
expect { doStuff() }.to(throwError(closure: closure))
```

I don't read that as an expectation on the type of error, but just on the fact that it should throw something that satisfy the given `closure`.

As such I am not expecting any information on the expected error type, because no such expectation has been set.

On the other hand something like this:

```swift
expect { throw NimbleError.laugh }.to(throwError(errorType: NimbleError.self, closure: closure))
```

I am expecting to have the error type information in the failure message.

Thoughts on this @norio-nomura @fbartho?

mokagio commented:
I implemented and pushed my idea from [the previous comment](https://github.com/norio-nomura/Nimble/pull/5#issuecomment-239262794), to better show how it looks like, and see if the tests pass: f22ff06.

I'm happy to remove that commit if a better solution is proposed.

mokagio commented:
[Travis' tests](https://travis-ci.org/norio-nomura/Nimble/builds/151611105) are failing 😔 Except for the OSX target which passes 🎉.

The failure on the iOS and tvOS targets looks to be happening on build time, while the SPM ones are due to:

```
error: the module at /Users/travis/build/norio-nomura/Nimble/Tests/Nimble contains mixed language source files
fix: use only a single language within a module
```

norio-nomura commented:
I pushed PRs.

mokagio commented:
@norio-nomura your PRs have been merged. Thanks.

norio-nomura commented:
@mokagio Thanks for merging my PRs! Now, all tests passed on Travis-CI. Do you have another works remaining on this?
If not, I'll merge this to the branch tracking https://github.com/Quick/Nimble/pull/277

mokagio commented:
Sweet. I don't have any work left. 🚢😊

Thanks for your help @norio-nomura getting this done.

norio-nomura commented:
Thanks! 🙏
